Support heapless environments (no_std without alloc)

dflemstr opened the pull request:
This is a big change that allows us to run the library without a heap (of course feature gated).

Some trade-offs that have been made:

  - I changed `Id` from using a pair of `u32` to using a single `u64`.  This removes the need to re-use IDs.  Conceivably we could overflow `u64` but this shouldn't happen until the death of the universe.
  - To get the current crate `no_std` + `alloc` behavior, you need to activate the `alloc` feature. I guess this is a breaking change if someone was relying on `default-features = false` in their `Cargo.toml`...
  - The public API no longer contains any allocations like `Vec` or `Box`, as a result measure functions can no longer return errors, and you need to use unboxed closures.  If there is a need it should be easy to add back boxed closures.
  - For the heapless mode, a cap had to be put on the number of elements supported.  The max number of nodes is 256 and only vecs up to size 8 are supported.  This can easily be controlled through the `sys` module.  One big limiting factor is that some algorithms use `Vec<Vec<...>>` which for heapless mode allocates 8*8=64 memory cells.  For non-heapless mode, we still support an indefinite number of nodes/elements.

dflemstr committed:
Make Style derive PartialEq

dflemstr committed:
Add ability to access child node by index

emilsjolander commented:
First of all, super thanks for this. Looks cool and was def a lot of work. Some comments below but i'm optimistic we can merge this. My main concern is around the iOS and Android crates which need to be updated. They currently build against published stretch version and not local version which is why the CI build is passing.

Let me know if I can help with those platforms in any way.

> I changed Id from using a pair of u32 to using a single u64. This removes the need to re-use IDs. Conceivably we could overflow u64 but this shouldn't happen until the death of the universe.

This sounds like a good change. 100% ok with this.

> To get the current crate no_std + alloc behavior, you need to activate the alloc feature. I guess this is a breaking change if someone was relying on `default-features = false` in their `Cargo.toml`

I'm OK with this, make a note of it in the documentation/readme? We are pre 1.0 for exactly this reason. Until more large-scale usage of the library is implemented we fully expect things like this to change.

> The public API no longer contains any allocations like Vec or Box, as a result measure functions can no longer return errors, and you need to use unboxed closures. If there is a need it should be easy to add back boxed closures.

This was mostly used to make interop with Swift and Kotlin easier (possible?). Because the currently iOS and Android crates rely on published stretch version as apposed to local version (I have been meaning to make this change) this breaking change is not reflected in the CircleCI runs for those projects. 

> For the heapless mode, a cap had to be put on the number of elements supported.

Can we make this a feature flag in `cargo.toml`? Or in some other way make it user-definable without forking the library?

dflemstr commented:
@emilsjolander thanks for the feedback, I also noticed that it's very easy to hit the limits of the stack-allocated buffers in algo.rs right now, so first I want to find a way to fix that, and then I will make sure that the Android and iOS builds work.

dflemstr commented:
(And I will also split the PR into commits that changes the prod code vs test and bench code because I guess it becomes hard to review now)

emilsjolander commented:
Thanks! That would be helpful indeed

dflemstr commented:
I'm now happy with the stack usage of the algorithm.  With the latest changes, and when running the `basic` example, the functions using the most stack space are:

```
0x0000000000000e80      344     main
0x0000000000025950      360     stretch::node::Stretch::new_node::hec1479ab44d2bc8e
0x0000000000001860      392     backtrace::print::BacktraceFrameFmt::print_raw::h4fb9cbf291fbf68a
0x0000000000022df0      408     stretch::forest::Forest::new_node::h47548da92cd23f9f
0x0000000000014a50      520     std::sys_common::backtrace::output_filename::h20848066042c7fe2
0x0000000000000370      552     basic::main::h1ca31bbb6fe9aeaf
0x000000000000d780      632     <rustc_demangle::v0::Ident as core::fmt::Display>::fmt::hc54ce657d4c948fb
0x000000000001bd60      888     stretch::algo::<impl stretch::forest::Forest>::compute_internal::h2dfe177d93cd7e9d
0x0000000000008990      1624    core::fmt::float::float_to_decimal_common_shortest::h63c768b484806025
0x000000000000a9f0      2008    core::fmt::float::float_to_decimal_common_exact::hbbd14c96ff317a43
```

Previously, `stretch::algo` was using way more.  I will now start breaking out the changes into nicer commits and maybe separate PRs.

dflemstr committed:
Break out a sys module that encapsulates target-specific APIs

This module should contain any types or code that is different depending
on whether we're compiling for std, alloc or just core.

dflemstr committed:
Make the core algorithms be allocation free

dflemstr committed:
Switch to ID generation using a single increasing u64

dflemstr committed:
Update all benchmarks to use the new API

dflemstr committed:
Update all tests to use the new API

dflemstr committed:
Update the basic example to use the new API

dflemstr committed:
Run tests for all supported feature combinations

dflemstr commented:
@emilsjolander I'm now happy with the core part of the library (well, as you said we can also add support for feature-controlled array sizes, but I'd want to explore that in a separate PR)

I'll try to figure out the bits for other platforms, feel free to provide any help/pointers if you can

dflemstr committed:
Fix js bindings

dflemstr committed:
Fix kotlin bindings

dflemstr committed:
Fix swift bindings

dflemstr committed:
Remove unused dependency

dflemstr committed:
No longer depend on lazy_static

dflemstr committed:
Disable arrayvec std feature by default

dflemstr committed:
Drop in support for no_std float operations

emilsjolander commented:
@dflemstr Let me know what you think the PR is in a good state to review. excited 🤩

dflemstr commented:
@emilsjolander I think we should be good to go now, I tried to separate out the test and benchmark changes into separate commits, but I'm not sure if GitHub has a good UX for reviewing while ignoring only those changes :(

dflemstr commented:
@emilsjolander bump!

emilsjolander commented:
Hey @dflemstr This looks great and I'm happy to merge :) Sorry for the long delay, we were in the middle of shipping a couple big new product updates and later this whole virus thing started happening. I expect to be a lot more responsive going forward.
